lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,663 @@
+		} else {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the input or output identified by the `serial_id` was not added by the sender
+			//    - the `serial_id` does not correspond to a currently added input
+			self.abort_negotiation(AbortReason::SerialIdUnknown)
+		}
+	}
+
+	fn receive_tx_add_output(mut self, serial_id: u64, output: TxOut) ->
+	Result<InteractiveTxStateMachine<Negotiating>, InteractiveTxStateMachine<NegotiationAborted>> {
+		// TODO: the sats amount is less than the dust_limit
+		self.context.received_tx_add_output_count += 1;
+		if self.context.received_tx_add_output_count > MAX_RECEIVED_TX_ADD_OUTPUT_COUNT {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//     - if has received 4096 `tx_add_output` messages during this negotiation
+			return self.abort_negotiation(AbortReason::ReceivedTooManyTxAddOutputs);
+		}
+
+		if output.value > MAX_MONEY {
+			// The receiving node:
+			// - MUST fail the negotiation if:
+			//		- the sats amount is greater than 2,100,000,000,000,000 (MAX_MONEY)
+			return self.abort_negotiation(AbortReason::ExceededMaximumSatsAllowed);
+		}
+
+		if let None = self.context.outputs.insert(serial_id, output) {
+			Ok(InteractiveTxStateMachine { context: self.context, state: Negotiating {} })
+		} else {
+			// The receiving node:
+			//  - MUST fail the negotiation if:
+			//    - the `serial_id` is already included in the transaction
+			self.abort_negotiation(AbortReason::DuplicateSerialId)
+		}
+	}
+
+	fn receive_tx_remove_output(mut self, serial_id: SerialId) ->
+		Result<InteractiveTxStateMachine<Negotiating>, InteractiveTxStateMachine<NegotiationAborted>> {
+		if let Some(output) = self.context.outputs.remove(&serial_id) {
+			Ok(InteractiveTxStateMachine { context: self.context, state: Negotiating {} })
+		} else {
+			self.abort_negotiation(AbortReason::SerialIdUnknown)
+		}
+	}
+
+	fn send_tx_add_input(mut self, serial_id: u64, input: TxIn) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.inputs.insert(serial_id, input);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_add_output(mut self, serial_id: SerialId, output: TxOut) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.insert(serial_id, output);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_input(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.inputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_remove_output(mut self, serial_id: SerialId) -> InteractiveTxStateMachine<Negotiating> {
+		self.context.outputs.remove(&serial_id);
+		InteractiveTxStateMachine { context: self.context, state: Negotiating {} }
+	}
+
+	fn send_tx_abort(mut self) -> InteractiveTxStateMachine<NegotiationAborted> {
```

Comment:
@dunxen not sure where we landed on this. is this something we want InteractiveTxConstructor to handle, or?

Reply:
Ah right, I think we will actually be better off having it here as we can check tx_signatures sent/received state to figure out if we can actually discard this negotiation:

https://github.com/lightning/bolts/blob/091397fc0798c4b2bdb36dbaa9f5b4b1d4a463e1/02-peer-protocol.md#requirements-7

Reply:
> MUST NOT have already transmitted tx_signatures

and I presume that the transmission of `tx_signatures` would be handled outside the interactive tx state machine?